service/src/main/java/org/broadinstitute/listener/relay/inspectors/TokenChecker.java

```diff
@@ -0,0 +1,61 @@
+package org.broadinstitute.listener.relay.inspectors;
+
+import com.auth0.jwt.JWT;
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.time.Instant;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TokenChecker {
+  private final HttpClient httpClient = HttpClient.newHttpClient();
+  private final String GOOGLE_OAUTH_SERVER = "https://www.googleapis.com/oauth2/v1/tokeninfo?access_token=";
+  private final Logger logger = LoggerFactory.getLogger(SamResourceClient.class);
+
+  public OauthInfo getOauthInfo(String token) throws IOException, InterruptedException {
+    var jwtExpiration = tryDecodeAsB2cToken(token);
+
+    var now = Instant.now();
+
+    if (jwtExpiration.isPresent()) {
+      if(jwtExpiration.get().isAfter(now))
+        return new OauthInfo(Optional.of(jwtExpiration.get()), "");
+      else
+        return new OauthInfo(Optional.empty(), "JWT expired");
+    } else {
+      var request = HttpRequest.newBuilder()
+          .uri(URI.create(GOOGLE_OAUTH_SERVER + token))
+          .build();
+
+      var oauthInfoResponse = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
+      var decoded = new Gson().fromJson(oauthInfoResponse.body(), GoogleOauthInfoResponse.class);
+
+      if(decoded.expires_in > 0)
+          return new OauthInfo(Optional.of(now.plusSeconds(decoded.expires_in)), decoded.error);
+      else
+        return new OauthInfo(Optional.empty(), decoded.error);
+    }
+  }
+
+  private Optional<Instant> tryDecodeAsB2cToken(String token) {
+    try {
+      var decoded = JWT.decode(token);
+      return Optional.of(decoded.getExpiresAt().toInstant());
+    } catch (com.auth0.jwt.exceptions.JWTDecodeException e) {
+      logger.info("Fail to check decode JWT", e);
+      return Optional.empty();
+    }
+  }
+}
+
+class GoogleOauthInfoResponse {
+  int expires_in;
```

Comment:
Even though this is mapped from a JSON object, I think camel case would work here, as Jackson (and Gson) map between snake and camel case when mapping JSON to a java object. A record could work here too, I think.
```suggestion
  int expiresIn;
```

Reply:
I'm gonna stick with `expires_in` since this class is really only meant for the api response from Google..it shouldn't be used anywhere else...I'd rather know what exactly Google returns instead of having to guess there's conversion happened by the json library

Reply:
OK. In any event I think it'd be better to declare it as a private inner class of `TokenChecker`, since no other class uses it.
```
  static private class GoogleOauthInfoResponse {
    int expires_in;
    String error;
  }
```

Reply:
cant make it private for some reason...will leave it as is

Reply:
That's odd, I tried this out on your branch and it worked fine for me. You can use `record` for this too:
```
  private record GoogleOauthInfoResponse(int expires_in, String error, String error_description) {}
```